### PR TITLE
Improve redo_log

### PR DESCRIPTION
**Pull request by lindig**
We have seen "Redo_log.RedoLogFailure("Cannot connect to redo log")" and it mostly stems from threads interfering. The redo_log module was not designed with thread-safety in mind. This path greatly reduces the problem. We might want to study the problem more closely but I would suggest to take this patch to make the problem smaller.

The key ingredient is taking the database lock before opening the redo log flushing the database to the redo_log. This blocks other operations.

### CODE REVIEW

**Comment by lindig** [comment]
@edwintorok @psafont could we review this? I believe it improves the situation short of re-writing the redo log implementation. See CP-45213

File: ocaml/database/redo_log.ml [hunk 1]
```diff
@@ -886,11 +886,12 @@ let flush_db_exn db log =
...
 
 let enable_and_flush db log reason =
-  enable_existing log reason ; flush_db_exn db log
+  enable_existing log reason ;
+  with_lock redo_log_creation_mutex @@ fun () -> flush_db_exn db log
 
 let enable_block_and_flush db log path =
   enable_block_existing log path ;
-  flush_db_exn db log
+  with_lock redo_log_creation_mutex @@ fun () -> flush_db_exn db log
```
Comment (psafont) [inline ocaml/database/redo_log.ml:894]: Why is this mutex held? I see it's only used for manipulating the Set of redo_logs

Comment (lindig) [reply ocaml/database/redo_log.ml:894]: I can test this. This one was suggested by @edwintorok and I tested it.

Comment (lindig) [reply ocaml/database/redo_log.ml:894]: Using `log.mutex` instead causes a deadlock.

**Review by psafont — APPROVED** [review]

File: ocaml/database/redo_log.ml (hunk 1 above)
Comment (edwintorok) [reply ocaml/database/redo_log.ml:894]: `with_active_redologs` holds this mutex and is what used to synchronize DB delta writes. Although there are some calls such as shutdown which are not always wrapped with that.

**Review by edwintorok — APPROVED** [review]
I think there are still race conditions between `delete` (or `shutdown`) and actively writing to the redo log, because not all callers of 'shutdown' or 'delete' call 'with_active_redologs'.
Doesn't matter if you only have 1 redo log (you'll be unsafe once they are shutdown anyway), but if you have multiple redo logs and only shut down 1 then you may get some exceptions raised.

It might be useful to acquire the per-log mutex too whenever performing an operation on it, but we need to be careful not to deadlock in that case, because callers aren't consistent about which locks they take.

OTOH forcibly shutting down a redolog may be useful (e.g. if its device is not able to complete IO operations successfully anymore).

Anyway I think this PR is an improvement and we should merge it, but there is more we could do.

**Pull request merged**